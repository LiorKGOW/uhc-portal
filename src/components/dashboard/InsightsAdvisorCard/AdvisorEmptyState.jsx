import React from 'react';

import {
  EmptyState,
  EmptyStateBody,
  EmptyStateHeader,
  EmptyStateIcon,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons/dist/esm/icons/search-icon';

import links from '../../../common/installLinks.mjs';
import ExternalLink from '../../common/ExternalLink';

const AdvisorEmptyState = () => (
  <EmptyState isFullHeight className="ocm-insights--advisor-empty-state">
    <EmptyStateHeader
      titleText="No Advisor recommendations"
      icon={<EmptyStateIcon icon={SearchIcon} />}
      headingLevel="h4"
    />
    <EmptyStateBody>
      This feature uses the Remote Health functionality of OpenShift Container Platform. For further
      details about Insights, see the{' '}
      <ExternalLink href={links.REMOTE_HEALTH_INSIGHTS}>OpenShift documentation</ExternalLink>.
    </EmptyStateBody>
  </EmptyState>
);

export default AdvisorEmptyState;
