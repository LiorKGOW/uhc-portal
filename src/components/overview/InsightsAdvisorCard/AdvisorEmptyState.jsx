import React from 'react';
import { SearchIcon } from '@patternfly/react-icons';
import {
  EmptyState, EmptyStateBody, EmptyStateIcon, Title,
} from '@patternfly/react-core';

import ExternalLink from '../../common/ExternalLink';

const AdvisorEmptyState = () => (
  <EmptyState isFullHeight className="ocm-insights--advisor-empty-state">
    <EmptyStateIcon icon={SearchIcon} />
    <Title size="lg" headingLevel="h4">
        No Advisor recommendations
    </Title>
    <EmptyStateBody>
        This feature uses the Remote Health functionality of
        OpenShift Container Platform. For further details about Insights, see the
      {' '}
      <ExternalLink href="https://docs.openshift.com/container-platform/latest/support/remote_health_monitoring/using-insights-to-identify-issues-with-your-cluster.html">
            OpenShift documentation
      </ExternalLink>
        .
    </EmptyStateBody>
  </EmptyState>
);

export default AdvisorEmptyState;
