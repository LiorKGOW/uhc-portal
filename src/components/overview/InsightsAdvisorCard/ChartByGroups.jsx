import React from 'react';
import PropTypes from 'prop-types';
import { Title, Flex, FlexItem } from '@patternfly/react-core';
import { ChartPie, ChartLegend } from '@patternfly/react-charts';
import { groupTagHitsByGroups } from '../overviewHelpers';

export const categoryMapping = {
  service_availability: 1,
  performance: 2,
  fault_tolerance: 3,
  security: 4,
};

/* PF's ChartLabel doesn't allow to create a link from the part of the label,
   thus, the custom title component is used */
const TitleComponent = ({
  data, index, x, y, style,
}) => {
  const entity = data[index];
  const { name, count, tags } = entity;

  return (
    <text x={x} y={y} style={style} dy={5}>
      <tspan x={x} y={y}>
        {name}
        :
        {' '}
        <a
          href={`${window.location.origin}/${APP_BETA ? 'beta/' : ''}openshift/insights/advisor/recommendations?category=${categoryMapping[tags]}`}
          className="enabled-link"
        >
          {count}
        </a>
      </tspan>
    </text>
  );
};

TitleComponent.propTypes = {
  data: PropTypes.array.isRequired,
  index: PropTypes.number.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  style: PropTypes.object.isRequired,
};

const ChartByGroups = ({ tagHits, groups }) => {
  const groupedRulesByGroups = groupTagHitsByGroups(tagHits, [...groups]);

  return (
    <Flex className="ocm-insights--groups-chart" direction={{ default: 'column' }}>
      <FlexItem>
        <Title size="lg" headingLevel="h2">
          Recommendations by category
        </Title>
      </FlexItem>
      <FlexItem>
        <ChartPie
          ariaTitle="Categories statistics"
          constrainToVisibleArea
          data={Object.entries(groupedRulesByGroups)
            .map(([title, group]) => ({ x: title, y: group.count }))}
          width={400}
          height={120}
          padding={{
            bottom: 30,
            left: 0,
            right: 300, // Adjusted to accommodate legend
            top: 10,
          }}
          labels={({ datum }) => `${datum.x}: ${datum.y}`}
          legendData={
            Object.entries(groupedRulesByGroups)
              .map(([title, { count, tags }]) => ({
                name: title, count, tags, symbol: { type: 'circle' },
              }))
          }
          legendPosition="right"
          legendAllowWrap
          legendComponent={(
            <ChartLegend
              labelComponent={<TitleComponent />}
              height={100}
              width={300}
              itemsPerRow={2}
              orientation="horizontal"
              className="pf-m-redhat-font"
              style={{ labels: { fontSize: 12 } }}
              gutter={30}
            />
          )}
        />
      </FlexItem>
    </Flex>
  );
};

ChartByGroups.propTypes = {
  tagHits: PropTypes.object.isRequired,
  groups: PropTypes.array.isRequired,
};

export default ChartByGroups;
