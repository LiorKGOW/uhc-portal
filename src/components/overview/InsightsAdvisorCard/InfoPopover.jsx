import React from 'react';
import { Popover } from '@patternfly/react-core';
import { HelpIcon } from '@patternfly/react-icons';
// eslint-disable-next-line camelcase
import { global_Color_dark_200 } from '@patternfly/react-tokens';

import ExternalLink from '../../common/ExternalLink';

const InfoPopover = () => (
  <Popover
    className="ocm-insights--info-popover"
    aria-label="What is Insights?"
    position="left"
    maxWidth="25rem"
    enableFlip
    bodyContent={(
      <>
        <p>
            Insights identifies and prioritizes risks to security, performance,
            availability, and stability of your clusters.
        </p>
        <p>
            This feature uses the Remote Health functionality of OpenShift Container Platform.
            For further details about Insights, see the
          {' '}
          <ExternalLink href="https://docs.openshift.com/container-platform/latest/support/remote_health_monitoring/using-insights-to-identify-issues-with-your-cluster.html">
            OpenShift documentation
          </ExternalLink>
          .
        </p>
      </>
              )}
  >
    <HelpIcon className="ocm-insights--risk-chart__popover-icon" color={global_Color_dark_200.value} />
  </Popover>
);

export default InfoPopover;
