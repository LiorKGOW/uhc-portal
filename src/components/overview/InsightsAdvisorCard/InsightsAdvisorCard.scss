$widget-min-height: 372px;
$groups-chart-max-width: 500px;

.ocm-insights {
  &--advisor-card {
    min-height: $widget-min-height;

    &__body {
      display: flex;
      flex-direction: column;

      .ocm-insights--advisor-empty-state {
        flex-grow: 1;
      }

      .ocm-insights--risk-chart {
        &__title {
          display: inline-block;
          margin-right: 9px;
        }

        &__popover-icon {
          &:hover {
            cursor: pointer;
            fill: var(--pf-global--Color--100);
            transition: ease 0.3s;
          }
        }

        .ocm-insights--risk-item {
          &__label {
            color: var(--pf-global--Color--200);
          }
        }

        &__items {
          padding-left: 1rem;
        }
      }

      .ocm-insights--advisor-notice {
        color: var(--pf-global--Color--200);

        &__icon {
          margin-right: 0.3rem;
        }
      }
    }
  }

  &--advisor-notice {
    margin-top: 0.6rem;
  }

  &--groups-chart {
    max-width: $groups-chart-max-width;
  }

  &--advisor-empty-state {
    .pf-c-empty-state__body {
      max-width: 550px;
    }
  }
}

.ocm-insights--info-popover {
  p:not(:last-child) {
    margin-bottom: 1rem;
  }
}
