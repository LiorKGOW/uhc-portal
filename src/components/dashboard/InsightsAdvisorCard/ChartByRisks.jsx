import React from 'react';
import PropTypes from 'prop-types';

import { Flex, FlexItem, Title } from '@patternfly/react-core';

import { advisorBaseName } from '~/common/getBaseName';

import { getSeverityName } from '../overviewHelpers';

import InfoPopover from './InfoPopover';

function ChartByRisks({ riskHits }) {
  return (
    <Flex className="ocm-insights--risk-chart" direction={{ default: 'column' }}>
      <FlexItem spacer={{ default: 'spacerLg' }}>
        <Title className="ocm-insights--risk-chart__title" size="lg" headingLevel="h2">
          Advisor recommendations by severity
        </Title>
        <InfoPopover />
      </FlexItem>
      <FlexItem className="ocm-insights--risk-chart__items" spacer={{ default: 'spacerLg' }}>
        <Flex
          justifyContent={{ default: 'justifyContentFlexStart' }}
          fullWidth={{ default: '50%' }}
          spaceItems={{
            sm: 'spaceItemsXl',
            md: 'spaceItemsLg',
            lg: 'spaceItems2xl',
            '2xl': 'spaceItems4xl',
          }}
        >
          {Object.entries({
            1: 0,
            2: 0,
            3: 0,
            4: 0,
            ...riskHits,
          })
            .reverse()
            .map(([riskNumber, count]) => (
              <FlexItem
                className="ocm-insights--items__risk-item"
                data-testid="ocm-insights--items__risk-item"
                key={riskNumber}
              >
                <Flex
                  direction={{ default: 'column' }}
                  alignItems={{ default: 'alignItemsCenter' }}
                  spaceItems={{ default: 'spaceItemsNone' }}
                >
                  <FlexItem className="ocm-insights--risk-item__count">
                    <Title size="2xl" headingLevel="h1">
                      <a href={`${advisorBaseName}/recommendations?total_risk=${riskNumber}`}>
                        {count}
                      </a>
                    </Title>
                  </FlexItem>
                  <FlexItem className="ocm-insights--risk-item__label">
                    {getSeverityName(riskNumber)}
                  </FlexItem>
                </Flex>
              </FlexItem>
            ))}
        </Flex>
      </FlexItem>
    </Flex>
  );
}

export default ChartByRisks;

ChartByRisks.propTypes = {
  riskHits: PropTypes.object.isRequired,
};
