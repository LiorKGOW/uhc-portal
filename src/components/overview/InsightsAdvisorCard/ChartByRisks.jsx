import React from 'react';
import PropTypes from 'prop-types';
import { Title, Flex, FlexItem } from '@patternfly/react-core';

import InfoPopover from './InfoPopover';
import { getSeverityName } from '../overviewHelpers';

function ChartByRisks({ riskHits }) {
  return (
    <Flex className="ocm-insights--risk-chart" direction={{ default: 'column' }}>
      <FlexItem spacer={{ default: 'spacerLg' }}>
        <Title className="ocm-insights--risk-chart__title" size="lg" headingLevel="h2">
          Advisor recommendations by severity
        </Title>
        <InfoPopover />
      </FlexItem>
      <FlexItem className="ocm-insights--risk-chart__items" spacer={{ default: 'spacerLg' }}>
        <Flex
          justifyContent={{ default: 'justifyContentFlexStart' }}
          fullWidth={{ default: '50%' }}
          spaceItems={{
            sm: 'spaceItemsXl', md: 'spaceItemsLg', lg: 'spaceItems2xl', '2xl': 'spaceItems4xl',
          }}
        >
          {Object.entries({
            1: 0,
            2: 0,
            3: 0,
            4: 0,
            ...riskHits,
          }).reverse().map(([riskNumber, count]) => (
            <FlexItem className="ocm-insights--items__risk-item" key={riskNumber}>
              <Flex
                direction={{ default: 'column' }}
                alignItems={{ default: 'alignItemsCenter' }}
                spaceItems={{ default: 'spaceItemsNone' }}
              >
                <FlexItem className="ocm-insights--risk-item__count">
                  <Title size="2xl" headingLevel="h1">
                    {
                      // TODO: Remove APP_BETA flag when OCP Advisor is in non-beta
                      APP_BETA
                        ? (
                          <a
                            href={`${window.location.origin}/${APP_BETA ? 'beta/' : ''}openshift/insights/advisor/recommendations?total_risk=${riskNumber}`}
                          >
                            {count}
                          </a>
                        )
                        : count
                    }
                  </Title>
                </FlexItem>
                <FlexItem className="ocm-insights--risk-item__label">
                  {getSeverityName(riskNumber)}
                </FlexItem>
              </Flex>
            </FlexItem>
          ))}
        </Flex>
      </FlexItem>
    </Flex>
  );
}

export default ChartByRisks;

ChartByRisks.propTypes = {
  riskHits: PropTypes.object.isRequired,
};
