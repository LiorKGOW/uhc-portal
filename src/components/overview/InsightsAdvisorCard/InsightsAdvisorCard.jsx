import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardBody } from '@patternfly/react-core';

import ChartByRisks from './ChartByRisks';
import ChartByGroups from './ChartByGroups';
import AdvisorEmptyState from './AdvisorEmptyState';
import { INSIGHTS_RULE_CATEGORIES } from '../../clusters/ClusterDetails/components/Insights/InsightsConstants';

import './InsightsAdvisorCard.scss';

const InsightsAdvisorCard = ({ overview }) => (
  <Card className="ocm-insights--advisor-card" ouiaId="insightsAdvisor">
    <CardBody className="ocm-insights--advisor-card__body">
      {!overview.clusters_hit || overview.clusters_hit === 0 ? (
        <AdvisorEmptyState />
      ) : (
        <>
          <ChartByRisks riskHits={overview.hit_by_risk} />
          <ChartByGroups tagHits={overview.hit_by_tag} groups={INSIGHTS_RULE_CATEGORIES} />
          <a
            href={`${window.location.origin}/${APP_BETA ? 'beta/' : ''}openshift/insights/advisor`}
            style={{ marginTop: '1rem' }}
          >
            View more in Insights Advisor
          </a>
        </>
      )}
    </CardBody>
  </Card>
);

export default InsightsAdvisorCard;

InsightsAdvisorCard.propTypes = {
  overview: PropTypes.object.isRequired,
};
