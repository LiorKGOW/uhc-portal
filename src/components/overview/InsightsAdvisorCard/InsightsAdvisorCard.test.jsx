import React from 'react';
import { shallow } from 'enzyme';
import InsightsAdvisorCard from './InsightsAdvisorCard';

const initialState = {
  overview: {
    clusters_hit: 0,
  },
  groups: [],
};

describe('<InsightsAdvisorCard />', () => {
  let wrapper;

  describe('When no Advisor recommendations (no rule hits)', () => {
    beforeAll(() => {
      wrapper = shallow(
        <InsightsAdvisorCard
          overview={initialState.overview}
          groups={initialState.groups}
        />,
      );
    });

    it('renders empty state', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('When recommendations are available (rule hits)', () => {
    beforeAll(() => {
      const overview = {
        clusters_hit: 3,
        hit_by_risk: {
          1: 1,
          2: 2,
        },
        hit_by_tag: {
          tag1: 3,
          tag2: 2,
          tag3: 2,
        },
      };
      const groups = [
        {
          title: 'title1',
          description: 'title1 desc',
          tags: ['tag1'],
        },
        {
          title: 'title2',
          description: 'title2 desc',
          tags: ['tag2'],
        },
      ];
      wrapper = shallow(
        <InsightsAdvisorCard
          overview={overview}
          groups={groups}
        />,
      );
    });

    it('renders two charts', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
