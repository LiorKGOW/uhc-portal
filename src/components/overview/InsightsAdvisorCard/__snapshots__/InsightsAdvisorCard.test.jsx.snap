// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InsightsAdvisorCard /> When no Advisor recommendations (no rule hits) renders empty state 1`] = `
<Card
  className="ocm-insights--advisor-card"
  ouiaId="insightsAdvisor"
>
  <CardBody
    className="ocm-insights--advisor-card__body"
  >
    <AdvisorEmptyState />
  </CardBody>
</Card>
`;

exports[`<InsightsAdvisorCard /> When recommendations are available (rule hits) renders two charts 1`] = `
<Card
  className="ocm-insights--advisor-card"
  ouiaId="insightsAdvisor"
>
  <CardBody
    className="ocm-insights--advisor-card__body"
  >
    <ChartByRisks
      riskHits={
        Object {
          "1": 1,
          "2": 2,
        }
      }
    />
    <ChartByGroups
      groups={
        Array [
          Object {
            "description": "title1 desc",
            "tags": Array [
              "tag1",
            ],
            "title": "title1",
          },
          Object {
            "description": "title2 desc",
            "tags": Array [
              "tag2",
            ],
            "title": "title2",
          },
        ]
      }
      tagHits={
        Object {
          "tag1": 3,
          "tag2": 2,
          "tag3": 2,
        }
      }
    />
    <div
      className="ocm-insights--advisor-notice"
    >
      <QuestionCircleIcon
        className="ocm-insights--advisor-notice__icon"
        color="#6a6e73"
        noVerticalAlign={false}
        size="sm"
      />
       
      Advisor information applies for OCP clusters only
    </div>
  </CardBody>
</Card>
`;
